protocol-specs/generators/multisignature_registration/index.js

```diff
@@ -14,164 +14,206 @@
-const SIGNATURE_PRESENT = Buffer.from('01', 'hex');
+const outputBase64Account = account => ({
+	...account,
+	privateKey: account.privateKey.toString('base64'),
+	publicKey: account.publicKey.toString('base64'),
+	address: account.address.toString('base64'),
+});
 
-const sortKeysAscending = publicKeys =>
-	publicKeys.sort((publicKeyA, publicKeyB) => {
-		// eslint-disable-next-line no-undef, new-cap
-		if (BigInt(`0x${publicKeyA}`) > BigInt(`0x${publicKeyB}`)) {
-			return 1;
-		}
-		// eslint-disable-next-line no-undef, new-cap
-		if (BigInt(`0x${publicKeyA}`) < BigInt(`0x${publicKeyB}`)) {
-			return -1;
-		}
-		return 0;
-	});
-
-const getId = transactionBytes => {
-	const transactionHash = hash(transactionBytes);
-	const transactionId = bufferToHex(transactionHash);
-
-	return transactionId;
+const baseSchema = {
+	$id: 'baseSchema',
+	type: 'object',
+	required: ['type', 'nonce', 'fee', 'senderPublicKey', 'asset'],
+	properties: {
+		type: {
+			dataType: 'uint32',
+			fieldNumber: 1,
+		},
+		nonce: {
+			dataType: 'uint64',
+			fieldNumber: 2,
+		},
+		fee: {
+			dataType: 'uint64',
+			fieldNumber: 3,
+		},
+		senderPublicKey: {
+			dataType: 'bytes',
+			fieldNumber: 4,
+		},
+		asset: {
+			dataType: 'bytes',
+			fieldNumber: 5,
+		},
+		signatures: {
+			type: 'array',
+			items: {
+				dataType: 'bytes',
+			},
+			fieldNumber: 6,
+		},
+	},
 };
 
-const assetToBytes = tx => {
-	const { mandatoryKeys, optionalKeys, numberOfSignatures } = tx.asset;
-	const mandatoryKeysBuffer = Buffer.from(mandatoryKeys.join(''), 'hex');
-	const optionalKeysBuffer = Buffer.from(optionalKeys.join(''), 'hex');
-
-	const assetBuffer = Buffer.concat([
-		intToBuffer(mandatoryKeys.length, 1),
-		mandatoryKeysBuffer,
-		intToBuffer(optionalKeys.length, 1),
-		optionalKeysBuffer,
-		intToBuffer(numberOfSignatures, 1),
-	]);
-	return assetBuffer;
+const multisigRegAsset = {
+	type: 'object',
+	properties: {
+		numberOfSignatures: { dataType: 'uint32', fieldNumber: 1 },
+		mandatoryKeys: {
+			type: 'array',
+			items: { dataType: 'bytes' },
+			fieldNumber: 2,
+		},
+		optionalKeys: {
+			type: 'array',
+			items: { dataType: 'bytes' },
+			fieldNumber: 3,
+		},
+	},
+	required: ['numberOfSignatures', 'mandatoryKeys', 'optionalKeys'],
 };
 
-const createSignatureObject = (txBuffer, account) => ({
-	signature: signData(
-		Buffer.concat([hexToBuffer(networkIdentifier), txBuffer]),
-		account.passphrase,
-	),
-});
-
-const serializeBasicProperties = tx => {
-	const transactionNonce = intToBuffer(tx.nonce.toString(), 8);
-
-	const buf = Buffer.concat([
-		Buffer.alloc(1, tx.type),
-		transactionNonce,
-		hexToBuffer(tx.senderPublicKey),
-		intToBuffer(tx.fee.toString(), 8),
-		assetToBytes(tx),
-	]);
+const getAssetBytes = asset => codec.encode(multisigRegAsset, asset);
 
-	return buf;
+const getSignBytes = tx => {
+	const assetBytes = getAssetBytes(tx.asset);
+	const signingTx = {
+		...tx,
+		asset: assetBytes,
+		signatures: [],
+	};
```

Comment:
Why can't get have single schema and single encode call?

Reply:
In base schema, asset is defined as bytes, and when we encode it has to be done in 2 steps, so it's also done in 2 steps as defined in LIP